### PR TITLE
Refactor transitions and Gaussian filters

### PR DESCRIPTION
**Pull request by pnkraemer**
I am back. Sorry for the confusion. These are the changes, take a look if you like. I am not done yet, but will do some more cleaning up.

### CODE REVIEW

**Comment by pnkraemer** [comment]
@nathanaelbosch

File: src/probnum/filtsmooth/statespace/sde.py
Comment (nathanaelbosch) [inline src/probnum/filtsmooth/statespace/sde.py:286]: Some simple error messages here and below would be great.

**Review by nathanaelbosch — CHANGES_REQUESTED** [review]
Alright! I did not yet review the tests, but I already have some feedback. Overall I really like where the code structure is going with this PR :)

File: src/probnum/diffeq/odefiltsmooth/ivpfiltsmooth.py
Comment (pnkraemer) [reply src/probnum/diffeq/odefiltsmooth/ivpfiltsmooth.py:122]: for me the tests break without. Is this just on my machine?

File: src/probnum/filtsmooth/gaussfiltsmooth/extendedkalman.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/extendedkalman.py:50]: for non-linear models this cannot be computed without evaluating functions (transitions are just some functions). A-priori, I dont know where to evaluate this, so maybe we can be clever about it in the future by reading it at the first `self.transition_rv` call, but I'd postpone such trickery to the future.

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:37]: Dont know. Maybe we try to track down kwargs. For now, I'd keep it at that^^

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:49]: they are there for interface reasons. The superclass has them, so I put them here for peer pressure reasons... Thoughts?

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:107]: Hmmm so far it inherits from `BayesFiltSmooth.predict` which is a bit of a meaningless docstring aka `Prediction step of the filter.`  Shall we keep it at that for now?

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:154]: Good question. I guess so, but not sure. Lets come back to this after the sampling PR is merged?

File: src/probnum/filtsmooth/statespace/discrete_transition.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/statespace/discrete_transition.py:121]: it is a relict from a past idea, before I decided to put those functions directly into the filters.

File: src/probnum/filtsmooth/statespace/sde.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/statespace/sde.py:69]: happy halloween ;) Just kidding, there is not.

File: src/probnum/filtsmooth/statespace/sde.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/statespace/sde.py:83]: not sure. I think I did it for adherence to interfaces. What do you think?

File: src/probnum/filtsmooth/statespace/transition.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/statespace/transition.py:155]: the additional stop argument. It could be swallowed by kwargs I guess, but would this be didactic enough?

File: src/probnum/filtsmooth/statespace/transition.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/statespace/transition.py:230]: because it requires evaluation at initrv, which is (in principle) decoupled from the transition.

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (nathanaelbosch) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:37]: I think in many functions the kwargs were just not documented, which I would personally prefer in order to keep the documentation clean.

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (nathanaelbosch) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:49]: I don't know about the best practices regarding function signatures of classes and subclasses. I guess they should be the same?

In any case, I'd be in favor of removing **kwargs wherever we don't explicitly need them. But I also understand that we might want to do that separately and not in this PR (although that seems like one of those issues that we never explicitly tackle), so it's also fine for me if you want to keep them in. If we postpone, it would be nice to open an issue for it.

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (nathanaelbosch) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:107]: Ok

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (nathanaelbosch) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:154]: Im not sure I completely understood you, do you mean that you will change that in the future, but still in this PR, or that you would like to postpone this for another PR? In case the sampling PR is merged while this one is still open, I think this change should be included in this PR.

File: src/probnum/filtsmooth/statespace/sde.py
Comment (nathanaelbosch) [reply src/probnum/filtsmooth/statespace/sde.py:82]: Im again not sure about what the best practices are. It sounds reasonable to me that functions in classes and subclasses should share function arguments, so that I'd be fine with keeping it the way it is. On the other hand, if "euler_step" is actually required, then I'd also like the code to reflect that, which is most easily and obviously done by having it as an explicit argument. 

Your call :)

File: tests/test_filtsmooth/test_gaussfiltsmooth/test_extendedkalman.py
Comment (nathanaelbosch) [inline tests/test_filtsmooth/test_gaussfiltsmooth/test_extendedkalman.py:21]: Should this file still be filled with tests for this PR?

File: tests/test_filtsmooth/test_statespace/test_discrete_transition.py
Comment (nathanaelbosch) [inline tests/test_filtsmooth/test_statespace/test_discrete_transition.py:144]: Should these stay?

**Review by nathanaelbosch — CHANGES_REQUESTED** [review]

File: tests/test_filtsmooth/test_gaussfiltsmooth/test_extendedkalman.py
Comment (pnkraemer) [reply tests/test_filtsmooth/test_gaussfiltsmooth/test_extendedkalman.py:21]: It inherits the tests from the superclass. This is, because UKF and EKF have the same tests.

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:37]: okay, lets do this.

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:49]: lets do this in a separate PR. I will create an issue.

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:154]: if the sampling PR is merged first, then of course I will get back to this one here. I am not sure though if this will happen... :)

File: src/probnum/filtsmooth/statespace/sde.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/statespace/sde.py:82]: I think you are right. Let's make explicit arguments in the "lowest" classes and capture those in the superclasses in the kwargs to maintain exchangability.

Comment (pnkraemer) [reply src/probnum/filtsmooth/statespace/sde.py:82]: screw my linter ;)

File: tests/test_filtsmooth/test_statespace/test_discrete_transition.py
Comment (pnkraemer) [reply tests/test_filtsmooth/test_statespace/test_discrete_transition.py:144]: no

File: src/probnum/filtsmooth/statespace/transition.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/statespace/transition.py:155]: I will remove it. What are the kwargs there for after all, if not for these cases...

**Comment by pnkraemer** [comment]
I made some minor improvements: linear_sde_statistics() has documentation now (and uses rk4 instead of euler). Also, the extended Kalman filter in continuous time is implemented.

File: src/probnum/filtsmooth/statespace/discrete_transition.py
Comment (nathanaelbosch) [inline src/probnum/filtsmooth/statespace/discrete_transition.py:39]: Why did you add a `**kwargs` here, was there some explicit reason? If yes we should keep track of that to understand if we want to have them around and when. 

And please don't get me wrong, I hope I'm not too annoying with those comments^^ If you think that it's best right now to have them around that's fine for me! I'm just a bit confused since I don't have the impression that the majority of our code even has them in function signatures, so I would personally refrain from adding new instances.

File: src/probnum/filtsmooth/statespace/sde.py
Comment (nathanaelbosch) [inline src/probnum/filtsmooth/statespace/sde.py:169]: Why is this needed? I suspect it's because `dispersionmatrix` is not a property in SDE, but it is here, right? Is this intended?

File: tests/test_filtsmooth/test_gaussfiltsmooth/filtsmooth_testcases.py
Comment (nathanaelbosch) [inline tests/test_filtsmooth/test_gaussfiltsmooth/filtsmooth_testcases.py:146]: Why are these required?

**Review by nathanaelbosch — CHANGES_REQUESTED** [review]
Looks good overall :) I had 3 more questions, which might not need any additional code changes. The only open thing seems to be the `final_rv` stuff, which should be adjusted in order to be up to date with the (soon to be?) merged sampling PR :)

File: src/probnum/filtsmooth/statespace/discrete_transition.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/statespace/discrete_transition.py:39]: I added them here, because they capture some of those extra arguments (e.g., `step`). Without, it is not possible to combine the two `generate` functions into one. 

No worries :) I think it is good to question those kwargs!

File: src/probnum/filtsmooth/statespace/sde.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/statespace/sde.py:169]: yes, this is an awkward spot in between a function that returns the dispersion at time $t$, though for the rest of the LTI SDE matrices they are properties.

File: tests/test_filtsmooth/test_gaussfiltsmooth/filtsmooth_testcases.py
Comment (pnkraemer) [reply tests/test_filtsmooth/test_gaussfiltsmooth/filtsmooth_testcases.py:146]: if I remember correctly, because the default values for `self.linearising_component` are `NotImplemented`, thus not callable

File: src/probnum/diffeq/odefiltsmooth/ivpfiltsmooth.py
Comment (nathanaelbosch) [reply src/probnum/diffeq/odefiltsmooth/ivpfiltsmooth.py:122]: I didn't try it, but if this is needed for tests to pass then it should be fine I guess.

File: src/probnum/filtsmooth/statespace/discrete_transition.py
Comment (nathanaelbosch) [reply src/probnum/filtsmooth/statespace/discrete_transition.py:39]: Cool, then of course it's totally appropriate to use it :)

**Review by nathanaelbosch — APPROVED** [review]
Looks good :) Thanks for your work!

**Pull request merged**
